Added dynamic context field w/ Autocomplete

arunapa opened the pull request:
Added a dynamic form to fill up metadata context (refer to screenshots)

Validated: 1) form submit, 2) if context is empty, form cannot be submit, 3) JSON output displayed on the final step

<img width="1624" alt="Screen Shot 2022-12-15 at 8 51 47 PM" src="https://user-images.githubusercontent.com/28983148/208011211-24d5c75e-a92a-4157-8a7e-cb2583c16648.png">

<img width="1624" alt="Screen Shot 2022-12-15 at 8 52 11 PM" src="https://user-images.githubusercontent.com/28983148/208011139-343dd030-6b3a-44a7-bf19-aec5eddd5870.png">

longshuicy commented:
I got an edge case error:
<img width="609" alt="image" src="https://user-images.githubusercontent.com/13950475/208018238-64453bbb-73d4-4193-9e15-44bba3dd0b52.png">

when i try to delete the last character of context this page errored out.
<img width="1160" alt="image" src="https://user-images.githubusercontent.com/13950475/208018356-d913b69c-1bf4-4e46-8786-b8c2386284fb.png">

If you watched console, it's complaining about autocomplete can't accept empty value and still cast it to lowercase. I'm suspecting the "freesolo" option need some condition.

arunapa commented:
> I got an edge case error: <img alt="image" width="609" src="https://user-images.githubusercontent.com/13950475/208018238-64453bbb-73d4-4193-9e15-44bba3dd0b52.png">
> 
> when i try to delete the last character of context this page errored out. <img alt="image" width="1160" src="https://user-images.githubusercontent.com/13950475/208018356-d913b69c-1bf4-4e46-8786-b8c2386284fb.png">
> 
> If you watched console, it's complaining about autocomplete can't accept empty value and still cast it to lowercase. I'm suspecting the "freesolo" option need some condition.

Thank you for catching this bug! I checked the documentation and after debugging, it looks like implementing both "onInputChange" and "onChange" was causing issues. When I  tried reproducing the autocomplete input, onInputChange event registered the new value as "" while onChange modified it to null, which is likely what the error was reporting.

I doubled checked the MUI API documentation and I think implementing onInputChange alone is sufficient.

longshuicy reviewed (approved):
Good finding about that onInputChange. It's working now. Approving!

lmarini reviewed (changes requested):
Can we rename to `Metadata context name` to `Term` and `Context URL` to `IRI`?

See [get_current_username](https://w3c.github.io/json-ld-syntax/#the-context) for more info.

Not on this PR, but in the future what if we moved the term mapping to each of the metadata entry. Currently I can create mappings in the context that don't match what is in the metadata definition. Ideally the Term will be the same used in the JSON for the metadata entry.

arunapa commented:
> Can we rename to `Metadata context name` to `Term` and `Context URL` to `IRI`?
> 
> See [get_current_username](https://w3c.github.io/json-ld-syntax/#the-context) for more info.
> 
> Not on this PR, but in the future what if we moved the term mapping to each of the metadata entry. Currently I can create mappings in the context that don't match what is in the metadata definition. Ideally the Term will be the same used in the JSON for the metadata entry.

Gotcha. I've renamed the fields both in my code and in the UI. 

Should we create an issue in the backlog for the future item?

<img width="1624" alt="Screen Shot 2022-12-16 at 11 23 01 AM" src="https://user-images.githubusercontent.com/28983148/208153910-4b8db427-cc1d-4728-b2ab-38f0c081db8e.png">

lmarini commented:
Yes, please. Thank you!

From: Aruna Parameswaran ***@***.***>
Date: Friday, December 16, 2022 at 11:23 AM
To: clowder-framework/clowder2 ***@***.***>
Cc: Marini, Luigi ***@***.***>, Review requested ***@***.***>
Subject: Re: [clowder-framework/clowder2] Added dynamic context field w/ Autocomplete (PR #244)

Can we rename to Metadata context name to Term and Context URL to IRI?

See get_current_username<https://urldefense.com/v3/__https:/w3c.github.io/json-ld-syntax/*the-context__;Iw!!DZ3fjg!_XsLt4l3vgNW80Yo0nPSFfLj7TTfok9HaZNLGhajbUMkdFSSGcnO7QjJZwpDHJ3ISjYuIKqfaakFPIXA61fOaa83Og$> for more info.

Not on this PR, but in the future what if we moved the term mapping to each of the metadata entry. Currently I can create mappings in the context that don't match what is in the metadata definition. Ideally the Term will be the same used in the JSON for the metadata entry.

Gotcha. I've renamed the fields both in my code and in the UI.

Should we create an issue in the backlog for the future item?

[Screen Shot 2022-12-16 at 11 23 01 AM]<https://urldefense.com/v3/__https:/user-images.githubusercontent.com/28983148/208153910-4b8db427-cc1d-4728-b2ab-38f0c081db8e.png__;!!DZ3fjg!_XsLt4l3vgNW80Yo0nPSFfLj7TTfok9HaZNLGhajbUMkdFSSGcnO7QjJZwpDHJ3ISjYuIKqfaakFPIXA61fqH3-IFQ$>

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/clowder-framework/clowder2/pull/244*issuecomment-1355260333__;Iw!!DZ3fjg!_XsLt4l3vgNW80Yo0nPSFfLj7TTfok9HaZNLGhajbUMkdFSSGcnO7QjJZwpDHJ3ISjYuIKqfaakFPIXA61e41CppXQ$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AAESNLIRZJ6CHZLB5QIJCQDWNSQRVANCNFSM6AAAAAAS72WCTI__;!!DZ3fjg!_XsLt4l3vgNW80Yo0nPSFfLj7TTfok9HaZNLGhajbUMkdFSSGcnO7QjJZwpDHJ3ISjYuIKqfaakFPIXA61dNUhgrLQ$>.
You are receiving this because your review was requested.Message ID: ***@***.***>

lmarini commented on frontend/src/metadata.config.tsx:
```diff
@@ -28,3 +28,32 @@ export const widgetTypes ={
     "DateTimePicker": "Date Time Picker",
     "Select": "Dropdown"
 }
+
```
Created https://github.com/clowder-framework/clowder2/issues/255. Hope it's not a duplicate.
